test-data/unit/check-overloading.test

```diff
@@ -1,5 +1,114 @@
 -- Test cases for function overloading
 
+[case testTypeCheckOverloadWithImplementation]
+from typing import overload, Any
+@overload
+def f(x: 'A') -> 'B': ...
+@overload
+def f(x: 'B') -> 'A': ...
+
+def f(x: Any) -> Any:
+   if isinstance(x, A):
+       return B()
+   else:
+       return A()
+
+reveal_type(f(A())) # E: Revealed type is '__main__.B'
+reveal_type(f(B())) # E: Revealed type is '__main__.A'
+
+class A: pass
+class B: pass
+[builtins fixtures/isinstance.pyi]
+
+[case testTypeCheckOverloadWithImplementationFastparse]
+# flags: --fast-parser
+
+from typing import overload, Any
+@overload
+def f(x: 'A') -> 'B': ...
+@overload
+def f(x: 'B') -> 'A': ...
+
+def f(x: Any) -> Any:
+   if isinstance(x, A):
+       return B()
+   else:
+       return A()
+
+reveal_type(f(A())) # E: Revealed type is '__main__.B'
+reveal_type(f(B())) # E: Revealed type is '__main__.A'
+
+class A: pass
+class B: pass
+[builtins fixtures/isinstance.pyi]
+
+[case testTypeCheckOverloadWithImplementationError]
+from typing import overload, Any
+
+@overload
+def f(x: 'A') -> 'B': ...
+@overload
+def f(x: 'B') -> 'A': ...
+
+def f(x: Any) -> Any:
+   if isinstance(x, A):
+       y = x
+       y = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
```

Comment:
Simplified the test and then re-complicated it by testing both the things you suggest.